core/server/src/main/java/alluxio/web/MasterUIWebServer.java

```diff
@@ -60,8 +63,23 @@ public MasterUIWebServer(ServiceType service, InetSocketAddress address, Alluxio
         "/header");
     mWebAppContext.addServlet(new ServletHolder(new WebInterfaceMasterMetricsServlet(
         master.getMasterMetricsSystem())), "/metricsui");
+
     // REST configuration
-    mWebAppContext.setOverrideDescriptors(Collections
-        .singletonList(Configuration.get(Constants.WEB_RESOURCES) + "/WEB-INF/master.xml"));
+    ResourceConfig config = new ResourceConfig().packages("alluxio.master", "alluxio.master.block",
+        "alluxio.master.file", "alluxio.master.lineage");
+    // Override the init method to inject a reference to AlluxioMaster into the servlet context.
+    // ServletContext may not be modified until after super.init() is called.
+    ServletContainer servlet = new ServletContainer(config) {
+      private static final long serialVersionUID = 7756010860672831556L;
+
+      @Override
+      public void init() throws ServletException {
+        super.init();
+        getServletContext().setAttribute(ALLUXIO_MASTER_SERVLET_RESOURCE_KEY, master);
+      }
+    };
+
+    ServletHolder servletHolder = new ServletHolder("Alluxio Master Web Service", servlet);
+    mWebAppContext.addServlet(servletHolder, "/v1/api/*");
```

Comment:
now that the REST configuration is all in Java (thanks a lot!) we should put `/v1/api` into constants as `REST_API_PREFIX` to make sure that the string constant is not duplicated anywhere